pkg/datastore/datastore.go

```diff
@@ -166,13 +206,49 @@ func (ds *memoryDatastore) RemoveOperatorSource(uid types.UID) {
 	delete(ds.rows, uid)
 }
 
-func (ds *memoryDatastore) GetOperatorSource(opsrcUID types.UID) (opsrc *v1alpha1.OperatorSourceSpec, ok bool) {
+func (ds *memoryDatastore) GetOperatorSource(opsrcUID types.UID) (*OperatorSourceKey, bool) {
 	row, exists := ds.rows[opsrcUID]
 	if !exists {
 		return nil, false
 	}
 
-	return row.Spec, true
+	return &row.OperatorSourceKey, true
+}
+
+func (ds *memoryDatastore) OperatorSourceHasUpdate(opsrcUID types.UID, metadata []*RegistryMetadata) (bool, error) {
+	// TODO: Return fine grained information that describes repository that
+	// was removed, added or has a new release.
+	row, exists := ds.rows[opsrcUID]
+	if !exists {
+		return false, fmt.Errorf("datastore has no record of the specified OperatorSource [%s]", opsrcUID)
+	}
+
+	// Right now we consider remote and local operator source to be same only
+	// when the following is true:
+	//  a. Number of repositories in both local and remote are exactly the same.
+	//  b. Each repository in remote has a corresponding local repository with
+	//     exactly the same release.
+
+	if len(row.Metadata) != len(metadata) {
+		return true, nil
+	}
+
+	for _, remote := range metadata {
+		if remote.Release == "" {
+			return false, fmt.Errorf("Release not specified for repository [%s]", remote.ID())
```

Comment:
- Should this be `false`? Won't it be safer to update the `OperatorSource` as not having a version signifies a change in the version? 
- What about the scenario where the repository never had a version in the first place?

Reply:
We should always have a release, I don't think we can push without specifying a release. I put it here as a defensive guard. `Not having a version` is treated as an error and hence I can't use it to indicate a change in version.